### ISSUE TITLE
DNSRR problem: not able to populate rdata on type=TXT records properly

### ISSUE THREAD

**Issue by emacsuser123** [secdev/scapy#1367]
#### Brief description

DNSRR stack does not allow rdata field to be populated properly when the type of record is TXT.

#### Environment

- Scapy version: 2.4.0
- Python version: 3.6
- Operating System: Ubuntu 17.10

#### How to reproduce

I'm trying to populate rdata in DNSRR (type=TXT), and it seems it only allows one 'subfield' to be populated.
The format for a TXT record is (as per it is being used):
(on rdlen field for the entire DNSRR record) and N of the below:
TXT length subfield 1
TXT subfield 1 (key1=value1)
TXT length subfield 2
TXT subfield 2 (key2=value2)
........
TXT length subfield N
TXT subfield N (keyN=valueN)
It seems that the rdata in scapy for TXT records only allows one of those subfields above.
I do not know how I can insert more than one subfield of the above.
When I insert the first subfield, it goes OK. It automatically inserts the subfield length associated, and
updates the global rdlen. But I have no way to insert additional subfields. I have tried to do it "manually", providing a string with all strings concatenated, including their lengths, but scapy then considers it as  only one subfield, and the auto-calculated subfield length does not reflect the real structure.
Maybe I'm missing something.

#### Actual result

I cannot provide more than one field to rdata, which is needed for TXT records.

#### Expected result

Being able to add more than one subfield to the rdata field.

Thanks and regards.
-Bob V

**Comment by guedou** [comment]
Please provide an example of code.

**Comment by emacsuser123** [comment]
Hello.

Ok.
I'm stacking three  additional resource records as part of the DNS reponse like this:

```
    # Answer section:
    packet[DNS].an = DNSRR(type='PTR', ttl=TTL_IN_RECORD)

    # Additional records section:
    ar1 = DNSRR(type='TXT', rclass=0x8001, ttl=TTL_TXT_IN_RECORD)
    ar2 = DNSRR(type='SRV', rclass=0x8001, ttl=TTL_IN_RECORD)
    ar3 = DNSRR(type='A', rclass=0x8001, ttl=TTL_IN_RECORD)

    # Stacking the three additional responses:
    packet[DNS].ar = ar1/ar2/ar3
```

Now, for the ar1 part, which is a TXT record, if I do, for instance:

`    packet[DNS].ar[0].rdata = '\x0agg=3242343\x10hh=kasdkfksadsad'
`
You can see that there are two subrecords:
gg=3242343
hh=kasdkfksadsad

each of them is prepended with their respective lengths byte. (0x0a = 10 bytes and 0x10 = 16 bytes)

When I send the packet out to the network, a wireshark(packet) statement shows ... (see attachment)

![screenshot from 2018-04-25 13-39-54](https://user-images.githubusercontent.com/16015875/39243593-426056aa-488e-11e8-9028-acdf28f4cf9a.png)


You can see in the selected section that the rdata field is not being considered as a "list" of pairs (length, field), but instead it is considered as one string. It puts the TXT length to 28 (total length). On top of that, you have the rdlen, which is OK.
I have tried different approaches: not including the length bytes, using a list of strings, ...

But none of them worked.

Not sure if somehow the RDataField class is not chopping the rdata field or something like that.

Thanks and best regards.
-Bob V

**Comment by guedou** [comment]
Providing data to the record works:
```
>>> dns = DNS()
>>> dns.an = DNSRR(type="TXT", rdata="hello")/ DNSRR(type="TXT", rdata="scapy")
>>> DNS(raw(dns))
<DNS  id=0 qr=0 opcode=QUERY aa=0 tc=0 rd=1 ra=0 z=0 ad=0 cd=0 rcode=ok qdcount=0 ancount=2 nscount=0 arcount=0 qd=None an=<DNSRR  rrname='.' type=TXT rclass=IN ttl=0 rdata='hello' |<DNSRR  rrname='.' type=TXT rclass=IN ttl=0 rdata='scapy' |>> ns=None ar=None |>
```
Did you try providing data to your records?

**Comment by guedou** [comment]
Can you point me to a RFC? I never heard of a DNS sub-record before.

**Comment by emacsuser123** [comment]
It was just my own slang  :-)

If you take a look at the wireshark snapshot attached, you will understand what I'm trying to say:
![dns](https://user-images.githubusercontent.com/16015875/39356680-f1b24cac-4a10-11e8-82a8-64f0fd09b701.png)

It is a DNS response with three additional responses. As per my understanding, that would be 3 DNSRR "instances". The first of them is a TXT record. The other two are SRV and A types, but we can leave them out of this discussion.
So, focusing on the TXT record:
The rdlength is 180.
What I call subrecord is:
- Subrecord 1:
Length: 35
key: id
value: fa1cc.....
- Subrecord2:
Length: 35
key: cd
value: 0FA ....

So, to sum up, there is only one DNSRR record for this TXT section. In its rdata field, there is a bunch of key=value pairs, with their own length, but that is internal to the field itself. One one rdlength is included.

When you do: DNSRR(type="TXT", rdata="hello")/ DNSRR(type="TXT", rdata="scapy")
Are you creating one record or two (each with its own rdlength)?

Thanks a lot.
-Bob V

**Comment by guedou** [comment]
Long story short: I learnt something today, and Scapy does not correctly support TXT records =)

I will try to provide a fix "soon". The DNS layer being one of the oldest onem it might take long to fix.

**Comment by emacsuser123** [comment]
Ok, no problem. Is there a quick way to somehow workaround it?
Something like passing a whole binary bunch of data to the enclosing DNSRR?
In fact that is one of the things I tried before and as soon as it sees type=TXT, the subrecord length is automatically added.
Not sure how to workaround it...

Thanks.

**Comment by guedou** [comment]
I can't thinkf of anything easier than patching the raw DNS packet.

**Comment by guedou** [comment]
@emacsuser123 I submitted a proposal to fix this issue in #1429

**Comment by emacsuser123** [comment]
Thanks a lot.

I will give it a try once it is available.

Regards.

2018-05-16 16:18 GMT+02:00 Guillaume Valadon <notifications@github.com>:
